src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -0,0 +1,777 @@
+    linearInterpolation(fragmentBarycentricCoords, getColumn(windowPoints, kZ)),
+    1 /
+      perspectiveInterpolation(
+        fragmentBarycentricCoords,
+        clipSpacePoints,
+        getColumn(clipSpacePoints, kW)
+      ),
+  ];
+}
+
+const kMultisamplingTables = new Map<number, readonly number[][]>([
+  [1, [[8 / 16, 8 / 16]]],
+  [
+    2,
+    [
+      [4 / 16, 4 / 16],
+      [12 / 16, 12 / 16],
+    ],
+  ],
+  [
+    4,
+    [
+      [6 / 16, 2 / 16],
+      [14 / 16, 6 / 16],
+      [2 / 16, 10 / 16],
+      [10 / 16, 14 / 16],
+    ],
+  ],
+  [
+    8,
+    [
+      [9 / 16, 5 / 16],
+      [7 / 16, 11 / 16],
+      [13 / 16, 9 / 16],
+      [5 / 16, 3 / 16],
+      [3 / 16, 13 / 16],
+      [1 / 16, 7 / 16],
+      [11 / 16, 15 / 16],
+      [15 / 16, 1 / 16],
+    ],
+  ],
+  [
+    16,
+    [
+      [9 / 16, 9 / 16],
+      [7 / 16, 5 / 16],
+      [5 / 16, 10 / 16],
+      [12 / 16, 7 / 16],
+
+      [3 / 16, 6 / 16],
+      [10 / 16, 13 / 16],
+      [13 / 16, 11 / 16],
+      [11 / 16, 3 / 16],
+
+      [6 / 16, 14 / 16],
+      [8 / 16, 1 / 16],
+      [4 / 16, 2 / 16],
+      [2 / 16, 12 / 16],
+
+      [0 / 16, 8 / 16],
+      [15 / 16, 4 / 16],
+      [14 / 16, 15 / 16],
+      [1 / 16, 0 / 16],
+    ],
+  ],
+]);
+
+/**
+ * Creates a function that will compute the interpolation of an inter-stage variable.
+ */
+function createInterStageInterpolationFn(
+  interStagePoints: number[][],
+  type: InterpolationType,
+  sampling: InterpolationSampling | undefined
+) {
+  return function ({
+    fragmentBarycentricCoords,
+    sampleBarycentricCoords,
+    clipSpacePoints,
+  }: FragData) {
+    const barycentricCoords =
+      sampling === 'center' ? fragmentBarycentricCoords : sampleBarycentricCoords;
+    switch (type) {
+      case 'perspective':
+        return interStagePoints[0].map((_, colNum: number) =>
+          perspectiveInterpolation(
+            barycentricCoords,
+            clipSpacePoints,
+            getColumn(interStagePoints, colNum)
+          )
+        );
+        break;
+      case 'linear':
+        return interStagePoints[0].map((_, colNum: number) =>
+          linearInterpolation(barycentricCoords, getColumn(interStagePoints, colNum))
+        );
+        break;
+      case 'flat':
+        return interStagePoints[0];
+        break;
+      default:
+        unreachable();
+    }
+  };
+}
+
+/**
+ * Renders float32 fragment shader inputs values to 4 rgba8unorm textures that
+ * can be multisampled textures. It stores each of the channels, r, g, b, a of
+ * the shader input to a separate texture, doing the math required to store the
+ * float32 value into an rgba8unorm texel.
+ *
+ * Note: We could try to store the output to an vec4f storage buffer.
+ * Unfortunately, using a storage buffer has the issue that we need to compute
+ * an index with the very thing we're trying to test. Similarly for a storage
+ * texture. Also, using a storage buffer seems to affect certain backends like
```

Comment:
nit. the "Similarly for storage texture" is a bit confusing? Maybe it was leftover comment?

Reply:
updated